File: src/algorithms/hash-maps/grouping/word-pattern/sources/word-pattern_test.rs
include!("word-pattern.rs");

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_returns_true_for_abba_dog_cat_cat_dog() {
        assert!(word_pattern("abba", "dog cat cat dog"));
    }

    #[test]
    fn test_returns_false_when_char_maps_to_two_words() {
        assert!(!word_pattern("abba", "dog cat cat fish"));
    }

    #[test]
    fn test_returns_true_for_aabb_dog_dog_cat_cat() {
        assert!(word_pattern("aabb", "dog dog cat cat"));
    }

    #[test]
    fn test_returns_false_when_all_same_but_pattern_varied() {
        assert!(!word_pattern("aaaa", "dog cat cat dog"));
    }

    #[test]
    fn test_returns_false_when_pattern_and_word_count_differ() {
        assert!(!word_pattern("abc", "dog cat"));
    }

    #[test]
    fn test_returns_true_for_single_char_pattern() {
        assert!(word_pattern("a", "dog"));
    }

    #[test]
    fn test_returns_true_for_identical_pattern_same_word() {
        assert!(word_pattern("aa", "dog dog"));
    }

    #[test]
    fn test_returns_false_when_bijection_violated_word_to_char() {
        assert!(!word_pattern("ab", "dog dog"));
    }

    #[test]
    fn test_handles_all_unique_chars_and_words() {
        assert!(word_pattern("abcd", "one two three four"));
    }
}

File: src/algorithms/hash-maps/grouping/word-pattern/__tests__/word-pattern_test.go
package main

import "testing"

func TestWordPattern_ReturnsTrueForAbbaDogCatCatDog(t *testing.T) {
	if !wordPattern("abba", "dog cat cat dog") {
		t.Error("expected true")
	}
}

func TestWordPattern_ReturnsFalseWhenCharMapsToTwoWords(t *testing.T) {
	if wordPattern("abba", "dog cat cat fish") {
		t.Error("expected false")
	}
}

func TestWordPattern_ReturnsTrueForAabbDogDogCatCat(t *testing.T) {
	if !wordPattern("aabb", "dog dog cat cat") {
		t.Error("expected true")
	}
}

func TestWordPattern_ReturnsFalseWhenAllSameButPatternVaried(t *testing.T) {
	if wordPattern("aaaa", "dog cat cat dog") {
		t.Error("expected false")
	}
}

func TestWordPattern_ReturnsFalseWhenPatternAndWordCountDiffer(t *testing.T) {
	if wordPattern("abc", "dog cat") {
		t.Error("expected false")
	}
}

func TestWordPattern_ReturnsTrueForSingleCharPattern(t *testing.T) {
	if !wordPattern("a", "dog") {
		t.Error("expected true")
	}
}

func TestWordPattern_ReturnsTrueForIdenticalPatternSameWord(t *testing.T) {
	if !wordPattern("aa", "dog dog") {
		t.Error("expected true")
	}
}

func TestWordPattern_ReturnsFalseWhenBijectionViolated(t *testing.T) {
	if wordPattern("ab", "dog dog") {
		t.Error("expected false")
	}
}

func TestWordPattern_HandlesAllUniqueCharsAndWords(t *testing.T) {
	if !wordPattern("abcd", "one two three four") {
		t.Error("expected true")
	}
}
